log/log.go

```diff
@@ -135,34 +146,50 @@ func (f LoggerFunc) Log(keyvals ...interface{}) {
 	f(keyvals...)
 }
 
+func getContext(l Logger) *Context {
+	if logger, ok := l.(LoggerWithContext); ok {
```

Comment:
The benchmark numbers for rate limited seem high.  I think this is returning NewContext when called under those circumstances. Can we not do the nil check i call out below? I feel like this is going to save us some calls. But the benchmarks will tell us.  Can you run that test with pprof and see where the allocations are that basically doubled?